pkg/controllers/externalsecret/externalsecret_controller.go

```diff
@@ -128,345 +154,465 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
+		syncCallsError.With(resourceLabels).Inc()
 		return ctrl.Result{}, err
 	}
 
-	// refresh should be skipped if
-	// 1. resource generation hasn't changed
-	// 2. refresh interval is 0
-	// 3. if we're still within refresh-interval
-	if !shouldRefresh(externalSecret) && isSecretValid(existingSecret) {
-		refreshInt = (externalSecret.Spec.RefreshInterval.Duration - timeSinceLastRefresh) + 5*time.Second
-		log.V(1).Info("skipping refresh", "rv", getResourceVersion(externalSecret), "nr", refreshInt.Seconds())
-		return ctrl.Result{RequeueAfter: refreshInt}, nil
+	// ensure the full cache is up-to-date
+	if !equality.Semantic.DeepEqual(secretPartial.GetObjectMeta(), existingSecret.GetObjectMeta()) {
+		log.V(1).Info("secret caches are not in sync, requeueing")
+		return ctrl.Result{Requeue: true}, nil
 	}
-	if !shouldReconcile(externalSecret) {
-		log.V(1).Info("stopping reconciling", "rv", getResourceVersion(externalSecret))
-		return ctrl.Result{}, nil
+
+	// refresh will be skipped if ALL the following conditions are met:
+	// 1. refresh interval is not 0
+	// 2. resource generation of the ExternalSecret has not changed
+	// 3. the last refresh time of the ExternalSecret is within the refresh interval
+	// 4. the target secret is valid:
+	//     - it exists
+	//     - it has the correct "managed" label
+	//     - it has the correct "data-hash" annotation
+	if !shouldRefresh(externalSecret) && isSecretValid(existingSecret) {
+		log.V(1).Info("skipping refresh")
+		return r.getRequeueResult(externalSecret), nil
 	}
 
-	// patch status when done processing
-	p := client.MergeFrom(externalSecret.DeepCopy())
+	// update status of the ExternalSecret when this function returns, if needed.
+	// we use the ability of deferred functions to update named return values `result` and `err`.
+	// NOTE: we dereference the DeepCopy of the status field because status fields are NOT pointers,
+	//       so otherwise the `equality.Semantic.DeepEqual` will always return false.
+	currentStatus := *externalSecret.Status.DeepCopy()
 	defer func() {
-		err = r.Status().Patch(ctx, &externalSecret, p)
-		if err != nil {
-			log.Error(err, errPatchStatus)
+		if !equality.Semantic.DeepEqual(currentStatus, externalSecret.Status) {
+			updateErr := r.Status().Update(ctx, externalSecret)
+			if updateErr != nil {
+				// if we got an update conflict, we should requeue immediately
+				if apierrors.IsConflict(updateErr) {
+					log.V(1).Info("conflict while updating status, will requeue")
+
+					// only update the result if there was no error, otherwise, we get an annoying log
+					// saying that results are ignored when there is an error, and errors are retried anyway.
+					if err == nil {
+						result = ctrl.Result{Requeue: true}
+					}
+				} else {
+					log.Error(updateErr, logErrorUpdateESStatus)
+					// only update the error if there is no error already
+					if err == nil {
+						err = updateErr
+					}
+				}
+			}
```

Comment:
Should most likely refactor this segmented into something more readable. It kills the complexity score. You are using the return `err` which is fine, but make it somehow more apparent please.

Reply:
@Skarlso I did have an explanation in the doc string above it:

```go
// update status of the ExternalSecret when this function returns, if needed.
// we use the ability of deferred functions to update named return values `result` and `err`.
// NOTE: we dereference the DeepCopy of the status field because status fields are NOT pointers,
//       so otherwise the `equality.Semantic.DeepEqual` will always return false.
```

Do you have a specific request for how to make it simpler in your view?

Reply:
I have flattened this function and added more docstrings in https://github.com/external-secrets/external-secrets/pull/4086/commits/2b8a4dd352a63c29dcccb8884ea2ee5a36f38374

Reply:
Thank you!